linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala

```diff
@@ -2016,29 +2024,43 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
             redo(CreateJSClass(className, newCaptureValues))(env)
           }
 
-        case _ =>
-          if (lhs == Lhs.Discard) {
-            /* Go "back" to transformStat() after having dived into
-             * expression statements. Remember that Lhs.Discard is a trick that
-             * we use to "add" all the code of pushLhsInto() to transformStat().
-             */
-            rhs match {
-              case _:Skip | _:VarDef | _:Assign | _:While |
-                  _:Debugger | _:JSSuperConstructorCall | _:JSDelete |
-                  _:StoreModule | Transient(_:SystemArrayCopy) =>
-                transformStat(rhs, tailPosLabels)
-              case _ =>
-                throw new IllegalArgumentException(
-                    "Illegal tree in JSDesugar.pushLhsInto():\n" +
-                    "lhs = " + lhs + "\n" +
-                    "rhs = " + rhs + " of class " + rhs.getClass)
-            }
-          } else {
-            throw new IllegalArgumentException(
-                "Illegal tree in JSDesugar.pushLhsInto():\n" +
-                "lhs = " + lhs + "\n" +
-                "rhs = " + rhs + " of class " + rhs.getClass)
+        // Statement-only trees
+
+        case _:Skip | _:VarDef | _:Assign | _:While | _:Debugger |
+            _:JSSuperConstructorCall | _:JSDelete | _:StoreModule |
+            Transient(_:SystemArrayCopy) =>
+          /* Go "back" to transformStat() after having dived into
+           * expression statements. This can only happen for Lhs.Discard and
+           * for Lhs.Return's whose target is a statement.
+           */
+          lhs match {
+            case Lhs.Discard =>
+              transformStat(rhs, tailPosLabels)
+            case Lhs.ReturnFromFunction =>
+              /* If we get here, it is because desugarToFunctionInternal()
+               * found a top-level Labeled and eliminated it. Therefore, unless
+               * we're mistaken, by construction we cannot be in tail position
+               * of the whole function. That means there is no point trying to
```

Comment:
```suggestion
               * of the whole function (otherwise doReturnToLabel would have
               * eliminated the lhs). That means there is no point trying to
```